jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -534,4 +543,128 @@ public Quadratic(int minCapacity, int maxCapacity, int maxBucketSize, long maxHe
             );
         }
     }
+
+    /**
+     * <p>A {@link Pool} implementation that uses a {@link ConcurrentPool} which overflows to a {@link QueuedPool}
+     * when its size grows over {@link ConcurrentPool#OPTIMAL_MAX_SIZE}.</p>
+     *
+     * @param <P> the type of the pooled objects
+     */
+    private static class CompoundPool<P> implements Pool<P>
+    {
+        private final Pool<P> primaryPool;
+        private final Pool<P> secondaryPool;
+
+        public CompoundPool(ConcurrentPool.StrategyType strategyType, int maxSize, boolean cache)
+        {
+            if (maxSize <= ConcurrentPool.OPTIMAL_MAX_SIZE)
+            {
+                this.primaryPool = new ConcurrentPool<>(strategyType, maxSize, cache);
+                this.secondaryPool = new NullPool<>();
+            }
+            else
+            {
+                this.primaryPool = new ConcurrentPool<>(strategyType, ConcurrentPool.OPTIMAL_MAX_SIZE, cache);
+                this.secondaryPool = new QueuedPool<>(maxSize - ConcurrentPool.OPTIMAL_MAX_SIZE);
```

Comment:
This logic has nothing to do with `CompoundPool` and everything to do with `ArrayByteBufferPool`.  It should be in the `ArrayByteBufferPool` constructor and this constructor should just take 2 pools.

Reply:
Done.